libs/langchain/langchain/vectorstores/elasticsearch.py

```diff
@@ -0,0 +1,1054 @@
+        **kwargs: Any,
+    ) -> List[str]:
+        """Run more texts through the embeddings and add to the vectorstore.
+
+        Args:
+            texts: Iterable of strings to add to the vectorstore.
+            metadatas: Optional list of metadatas associated with the texts.
+            ids: Optional list of ids to associate with the texts.
+            refresh_indices: Whether to refresh the Elasticsearch indices
+                            after adding the texts.
+            create_index_if_not_exists: Whether to create the Elasticsearch
+                                        index if it doesn't already exist.
+
+        Returns:
+            List of ids from adding the texts into the vectorstore.
+        """
+        try:
+            from elasticsearch.helpers import bulk
+        except ImportError:
+            raise ImportError(
+                "Could not import elasticsearch python package. "
+                "Please install it with `pip install elasticsearch`."
+            )
+
+        embeddings = []
+        ids = ids or [str(uuid.uuid4()) for _ in texts]
+        requests = []
+
+        if self.embedding is not None:
+            # If no search_type requires inference, we use the provided
+            # embedding function to embed the texts.
+            embeddings = self.embedding.embed_documents(list(texts))
+            dims_length = len(embeddings[0])
+
+            if create_index_if_not_exists:
+                self._create_index_if_not_exists(
+                    index_name=self.index_name, dims_length=dims_length
+                )
+
+            for i, (text, vector) in enumerate(zip(texts, embeddings)):
+                metadata = metadatas[i] if metadatas else {}
+
+                requests.append(
+                    {
+                        "_op_type": "index",
+                        "_index": self.index_name,
+                        self.query_field: text,
+                        self.vector_query_field: vector,
+                        "metadata": metadata,
+                        "_id": ids[i],
+                    }
+                )
+
+        else:
+            # the search_type doesn't require inference, so we don't need to
+            # embed the texts.
+            if create_index_if_not_exists:
+                self._create_index_if_not_exists(index_name=self.index_name)
+
+            for i, text in enumerate(texts):
+                metadata = metadatas[i] if metadatas else {}
+
+                requests.append(
+                    {
+                        "_op_type": "index",
+                        "_index": self.index_name,
+                        self.query_field: text,
+                        "metadata": metadata,
+                        "_id": ids[i],
+                    }
+                )
+
+        if len(requests) > 0:
+            try:
+                success, failed = bulk(self.client, requests, stats_only=True)
```

Comment:
You can add `refresh=refresh_indices` in the parameter to avoid calling refresh explicitly below. It should be more performant since only the shards that receive updates will apply the refresh.

Reply:
Yes nice little performance improvement! Done